src/kernel/heap.zig

```diff
@@ -0,0 +1,388 @@
+
+        // The end of the allocation is marked with a 0 bit so search for the requested size plus one extra block for the 0
+        if (self.search(self.start, size, self.size, alignment)) |result| {
+            var i: u32 = 0;
+            // Set the found entries as allocated
+            while (i < result.entries) : (i += 1) {
+                // Set the entry as allocated
+                // Cannot error as the entry being set will not be outside of the heap
+                self.bitmap.setEntry(@intCast(u32, result.entry + i)) catch unreachable;
+            }
+            return result.addr;
+        }
+        return null;
+    }
+
+    ///
+    /// Free previously allocated memory. It is recommended to not call this directly and instead use the Allocator interface.
+    ///
+    /// Arguments:
+    ///     INOUT self: *Heap - The heap to free within
+    ///     IN ptr: usize - The address of the allocation to free. Should have been returned from a prior call to alloc.
+    ///     IN len: usize - The size of the allocated region.
+    ///
+    /// Error: Heap.Error.
+    ///     Heap.Error.NotAllocated: The address hasn't been allocated or is outside of the heap.
+    ///
+    pub fn free(self: *Self, ptr: usize, len: usize) Error!void {
+        if (ptr < self.start or ptr + len > self.start + self.size)
+            return Error.NotAllocated;
+
+        const addr = ptr - self.start;
+        const addr_end = addr + len;
+
+        var addr_entry = @intCast(u32, addr / self.block_min_size);
+        // Make sure the entry for this address has been allocated
+        // Won't error as we've already checked the address is valid above
+        if (!(self.bitmap.isSet(addr_entry) catch unreachable))
+            return Error.NotAllocated;
+
+        const NodeSearch = struct {
+            start: usize,
+            end: usize,
+
+            const Self2 = @This();
+
+            pub fn search(min: usize, max: usize, order_min: usize, order_max: usize) ?Self2 {
+                if (min == order_min and max == order_max)
+                    return Self2{ .start = order_min, .end = order_max };
+                if (order_min > min or order_max < max)
+                    return null;
+                const order_size = order_max - order_min;
+                if (search(min, max, order_min, order_min + order_size / 2) orelse search(min, max, order_max - order_size / 2, order_max)) |r|
+                    return r;
+                return Self2{ .start = order_min, .end = order_max };
+            }
+        };
+
+        // Since the address could be aligned it may not fall on an allocation boundary, so it's necessary to traverse the tree to find the smallest node in which the address range fits
+        // This will not be null as the address range is already guaranteed to be within the heap and so at least the root node will fit it
+        const search_result = NodeSearch.search(addr, addr_end, 0, self.size) orelse unreachable;
```

Comment:
If it is guaranteed, then returning null isn't needed and could then replace the return null with `unreachable` in the search function. So then won't need the orelse.

Reply:
The search function is recursive and so it needs to know when a recursive call has failed. Only the root call is guaranteed to succeed.